### ISSUE TITLE
training steps in ttt

### ISSUE THREAD

**Issue by liren-jin** [a1600012888/LaCT#11]
hi, thanks for releasing the code of such great work! It is fascinating.
i have one question regarding the update setup in ttt. 
It seems that the ttt layer (W0 W1 W2) is only updated with one step. Would it be helpful to use more iterations in the inner level to get better updated weights?

**Comment by a1600012888** [comment]
Great question!

Yes, in the current setup, the TTT update is performed with a single iteration per chunk. In the autoregressive (AR) setting where there are multiple chunks, the fast weights are updated once per chunk, so effectively multiple updates happen across the sequence—but still only one update per chunk.

Your question touches on an interesting research direction. It's connected to test-time scaling, recurrent depth, and "the utilization rate" of the memory modules, etc. 

In the NVS setup, we did some early experiments with multi-step updates—specifically using two or four iterations. Unfortunately, we didn't observe improvements: two steps performed similarly or slightly worse than one step, and more steps led to unstable training. That said, I believe there's a lot of room for research in making multi-step updates stable and effective. I’m optimistic that with the right techniques, it can work.


A related piece of work is the paper ["Just Read Twice: Closing the Recall Gap for Recurrent Language Models"](https://arxiv.org/abs/2310.01405). It feeds the same context twice, which is conceptually similar to updating the fast weights multiple times "with the same set of tokens". The paper shows performance improvements. 
However, it's not exactly identical: repeating tokens along the sequence dimension causes different key-value pairs during the second pass, defining a different TTT objective when perform ttt-update for the repeated tokens.

**Comment by a1600012888** [comment]
I think this is an important research question. 

> will the previous information be largely forgotten? 

For the Single niddle in a haystack results, you can see that the model can successfully recall some results but not all. 

So, I think every update will forget something. But it's hard to measure: how much being remembered and how much being forgotten. 

And you need to find a way to measure so that to continue research along this direction. 


> randomly adding historical chunks in the batch used for TTT updates?

Maybe some architecture stuff and some ttt-training related stuff might help. Maybe will the correct training data, you don't need to do any design.   Some thing pops in my head for this, one is the ttt-training objective used in the Mesanet: https://arxiv.org/abs/2506.05233
the concept of locally optimal, defines a new ttt loss, which intuitively has less forgetting. (but might also has less memorization of new concept). 


I think eventually, it all comes to the state-size and flops used to update/query the memory. So for me, the most important thing at the architecture level to improve the memory is to find a better recipe (ttt loss, ttt update, ttt function etc.) with hardware efficiency, so that you can further scaling it.

**Comment by liren-jin** [comment]
Thanks for sharing these insights!
